### PR TITLE
feat: 4 add criteria query for books(title, author)

### PR DESCRIPTION
**Pull request by EduarfF**

### CODE REVIEW

File: src/main/java/org/example/springbootintro/repository/book/BookSpecificationBuilder.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import org.example.springbootintro.repository.SpecificationBuilder;
+import org.example.springbootintro.repository.SpecificationProviderManager;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
+    @Autowired
+    private SpecificationProviderManager<Book> bookSpecificationProviderManager;
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/springbootintro/repository/book/BookSpecificationBuilder.java]: let's pass it as a constructor parameter

File: src/main/java/org/example/springbootintro/repository/book/spec/AuthorSpecificationProvider.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import org.example.springbootintro.model.Book;
+import org.example.springbootintro.repository.SpecificationProvider;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "author";
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/springbootintro/repository/book/spec/AuthorSpecificationProvider.java]: let's make it a `public` constant

File: src/main/java/org/example/springbootintro/repository/book/BookSpecificationBuilder.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    public Specification<Book> build(BookSearchParametersDto searchParametersDto) {
+        Specification<Book> spec = Specification.where(null);
+        if (searchParametersDto.title() != null && searchParametersDto.title().length > 0) {
+            spec = spec.and(bookSpecificationProviderManager
+                    .getSpecificationProvider("title")
+                    .getSpecification(searchParametersDto.title()));
+        }
+        if (searchParametersDto.author() != null && searchParametersDto.author().length > 0) {
+            spec = spec.and(bookSpecificationProviderManager
+                    .getSpecificationProvider("author")
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/springbootintro/repository/book/BookSpecificationBuilder.java]: and reuse here as well

**Review by yevheniisynytsia — CHANGES_REQUESTED** [review]
Well done! Almost there!

File: src/main/java/org/example/springbootintro/repository/book/BookSpecificationBuilder.java [hunk 4]
```diff
@@ -4,14 +4,18 @@
...
 public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
-    @Autowired
-    private SpecificationProviderManager<Book> bookSpecificationProviderManager;
+    private final SpecificationProviderManager<Book> bookSpecificationProviderManager;
+
+    public BookSpecificationBuilder(
+            SpecificationProviderManager<Book> bookSpecificationProviderManager
+    ) {
+        this.bookSpecificationProviderManager = bookSpecificationProviderManager;
+    }
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/springbootintro/repository/book/BookSpecificationBuilder.java]: use lombok please

File: src/main/java/org/example/springbootintro/repository/book/BookSpecificationBuilder.java (hunk 3 above)
Comment (yevheniisynytsia) [reply src/main/java/org/example/springbootintro/repository/book/BookSpecificationBuilder.java]: please make your constant from AuthorSpecificationProvider public and use it here

File: src/main/java/org/example/springbootintro/repository/book/spec/TitleSpecificationProvider.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+import org.example.springbootintro.model.Book;
+import org.example.springbootintro.repository.SpecificationProvider;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class TitleSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "title";
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/springbootintro/repository/book/spec/TitleSpecificationProvider.java]: please add constant here as well

**Review by yevheniisynytsia — CHANGES_REQUESTED** [review]

**Review by karma-o — APPROVED** [review]
Nice!

**Pull request merged**